### PR TITLE
Reduce entity tracker updates on move

### PR DESCRIPTION
**Pull request by froobynooby**
**See patch notes for full description**

This patch reduces the number of unnecessary entity tracker updates that occur when movement packets are processed by the server. In testing, it has reduced the time taken to process such packets by up to 4x. Packet processing for movement packets often show high in timings reports and are usually the cause for oversleep

Although I have tested this on my own and on my small survival server (and noticed no issues), I haven't got the resources to test this on servers with any more than about 10 concurrent players. Although it _should_ scale well with player count, external testing would be appreciated

Paperclip (based on 112): https://drive.google.com/file/d/1j2hSTpOFbGo8ioeOM7vZal9wAgzUWWra/view

### CODE REVIEW

**Comment by HexedHero** [comment]
Gods work

**Comment by SrBedrock** [comment]
So that I can test with a base of 50~80 players, would it be possible to make the .jar available?

**Comment by kickash32** [comment]
Just based on the description of the patch, I think this patch is unsafe. What happens when an entity moves towards a player's view distance from outside of it? If the player has not moved, for one reason or another, the entity will never be visible until the player decides to move. Or what happens when a mobs spawns and attacks the player? I would like to know how this patch handles these edge cases.

**Comment by brodybeckwith** [comment]
> 
> 
> So that I can test with a base of 50~80 players, would it be possible to make the .jar available?

You should be able to just clone his repository and do `./paper jar` like you would for Paper regularly and it will generate the jar.

**Comment by froobynooby** [comment]
> So that I can test with a base of 50~80 players, would it be possible to make the .jar available?

Added a paperclip download to the main post

> Just based on the description of the patch, I think this patch is unsafe. What happens when an entity moves towards a player's view distance from outside of it? If the player has not moved, for one reason or another, the entity will never be visible until the player decides to move. Or what happens when a mobs spawns and attacks the player? I would like to know how this patch handles these edge cases.

When a player is standing still the behaviour of the tracker updates is slightly different, but what you describe should not be a problem
**Standing still without this patch**: client sends its position to the server once a second, when the server receives this it updates all of the trackers.
**Standing still with this patch**: client sends it position to the server once a second, when the server receives this it checks if it has been 20 ticks since last time and, if so, updates all the trackers.

So if a mob moves into the tracking range of a player who is standing still, with this patch it will take on average 20 and up to a maximum of 40 ticks before the player knows about it.

If the mob spawns then when it is added to the PlayerChunkMap the player will be instantly made aware of it, this is something the patch doesn't change

**Comment by HexedHero** [comment]
> So that I can test with a base of 50~80 players, would it be possible to make the .jar available?

Any results so far?

**Comment by Fejm** [comment]
Tested with 100 players online. TPS before patch: ~9.5, TPS after patch: ~15.5

**Comment by HexedHero** [comment]
Only a few hours of testing but with 30~ players it definitely improved my avg tick time with no side effects seen

**Comment by SrBedrock** [comment]
I'm not sure if the error is caused by the patch but I think I should report it.

```
>.... [16:13:26 FATAL]: Error executing task on Server
java.lang.NullPointerException: null
at net.minecraft.server.v1_15_R1.PlayerChunkMap$EntityTracker.updatePlayer(PlayerChunkMap.java:1770) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.PlayerChunkMap.movePlayer(PlayerChunkMap.java:1383) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.ChunkProviderServer.movePlayer(ChunkProviderServer.java:811) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.PlayerConnection.a(PlayerConnection.java:496) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.PacketPlayInTeleportAccept.a(SourceFile:30) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.PacketPlayInTeleportAccept.a(SourceFile:8) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.PlayerConnectionUtils.lambda$ensureMainThread$0(PlayerConnectionUtils.java:23) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.TickTask.run(SourceFile:18) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.executeTask(IAsyncTaskHandler.java:136) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.executeNext(IAsyncTaskHandler.java:109) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.MinecraftServer.ba(MinecraftServer.java:1038) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.MinecraftServer.executeNext(MinecraftServer.java:1031) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.awaitTasks(IAsyncTaskHandler.java:119) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.MinecraftServer.a(MinecraftServer.java:1102) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at net.minecraft.server.v1_15_R1.MinecraftServer.run(MinecraftServer.java:934) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
```
Using for the second day with an average of 70 players, I removed the [EntityTrackerFixer](https://www.spigotmc.org/resources/entitytrackerfixer-fix-1-14-4-1-15-entitytick-lag.70902/updates) plugin so there would be no problem. Performance has improved significantly.

**Comment by HexedHero** [comment]
> I'm not sure if the error is caused by the patch but I think I should report it.
> 
> ```
> >.... [16:13:26 FATAL]: Error executing task on Server
> java.lang.NullPointerException: null
> at net.minecraft.server.v1_15_R1.PlayerChunkMap$EntityTracker.updatePlayer(PlayerChunkMap.java:1770) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.PlayerChunkMap.movePlayer(PlayerChunkMap.java:1383) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.ChunkProviderServer.movePlayer(ChunkProviderServer.java:811) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.PlayerConnection.a(PlayerConnection.java:496) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.PacketPlayInTeleportAccept.a(SourceFile:30) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.PacketPlayInTeleportAccept.a(SourceFile:8) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.PlayerConnectionUtils.lambda$ensureMainThread$0(PlayerConnectionUtils.java:23) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.TickTask.run(SourceFile:18) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.executeTask(IAsyncTaskHandler.java:136) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.executeNext(IAsyncTaskHandler.java:109) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.MinecraftServer.ba(MinecraftServer.java:1038) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.MinecraftServer.executeNext(MinecraftServer.java:1031) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.awaitTasks(IAsyncTaskHandler.java:119) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.MinecraftServer.a(MinecraftServer.java:1102) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at net.minecraft.server.v1_15_R1.MinecraftServer.run(MinecraftServer.java:934) ~[patched_1.15.2.jar:git-Paper-"49495b62"]
> at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
> ```
> 
> Using for the second day with an average of 70 players, I removed the [EntityTrackerFixer](https://www.spigotmc.org/resources/entitytrackerfixer-fix-1-14-4-1-15-entitytick-lag.70902/updates) plugin so there would be no problem. Performance has improved significantly.

Its https://github.com/PaperMC/Paper/issues/2966

**Comment by HexedHero** [comment]
Sorry its https://github.com/PaperMC/Paper/issues/2803

**Comment by froobynooby** [comment]
> I'm not sure if the error is caused by the patch but I think I should report it.

The error is happening on a line the patch added, however the reason it is happening is that movement logic is being run either before the player has been added to the world, or after the player has been removed from the world. I made a little test plugin that would call movePlayer on a world I was _not_ in and I was able to trigger the same error as in 2803 quite consistently just by going through a nether portal. So I think Hexed is right - it is the same underlying issue as in 2803, it looks a little different with this patch

**Comment by SrBedrock** [comment]
New error/crash:

- Full log: https://hasteb.in/ipevozif.md
- Error: https://hasteb.in/kayuduha.php
- Crash Report: https://hasteb.in/oqojurom.php

**Comment by Spottedleaf** [comment]
The exception is not at all relating to 2803

**Comment by froobynooby** [comment]
> The exception is not at all relating to 2803

I made the following two assumptions when making this patch:

- movePlayer would never be called before the player was added to PlayerChunkMap with addEntity
- movePlayer would never be called after the player was removed from PlayerChunkMap with removeEntity

When I say it is the same underlying issue as 2803 I mean to say if the two conditions above were always met, neither the exception in 2803 or the exception he's getting with this patch could happen.

I'll add some null checks and the exceptions / crashes should go away

**Comment by SrBedrock** [comment]
Post the link here after the changes so I can continue testing.

**Comment by BlueSoapTurtle** [comment]
I tried out this patch and citizens disappeared, anyone else able to confirm this?

**Comment by froobynooby** [comment]
Paperclip download has been updated


> I tried out this patch and citizens disappeared, anyone else able to confirm this?

Are you able to provide steps for reproducing this? I've tried spawning in NPCs with Citizens and haven't had any issues with them

**Comment by SrBedrock** [comment]
After staying several hours AFK.
I don't know if it's related to the patch.

```
[04:55:48] [Paper Watchdog Thread/ERROR]: --- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH  - git-Paper-"9f5fadcc" (MC: 1.15.2) ---
[04:55:48] [Paper Watchdog Thread/ERROR]: The server has not responded for 13 seconds! Creating thread dump
[04:55:48] [Paper Watchdog Thread/ERROR]: ------------------------------
[04:55:48] [Paper Watchdog Thread/ERROR]: Server thread dump (Look for plugins here before reporting to Paper!):
[04:55:48] [Paper Watchdog Thread/ERROR]: ------------------------------
[04:55:48] [Paper Watchdog Thread/ERROR]: Current Thread: Server thread
[04:55:48] [Paper Watchdog Thread/ERROR]: 	PID: 22 | Suspended: false | Native: false | State: RUNNABLE
[04:55:48] [Paper Watchdog Thread/ERROR]: 	Stack:
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.EntityInsentient.cj(EntityInsentient.java:1244)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.EntityLiving.movementTick(EntityLiving.java:2546)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.EntityInsentient.movementTick(EntityInsentient.java:517)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.EntityAgeable.movementTick(EntityAgeable.java:171)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.EntityAnimal.movementTick(EntityAnimal.java:28)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.EntityLiving.tick(EntityLiving.java:2383)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.EntityInsentient.tick(EntityInsentient.java:286)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.WorldServer.entityJoinedWorld(WorldServer.java:759)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.WorldServer$$Lambda$3705/949571267.accept(Unknown Source)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.World.a(World.java:858)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.WorldServer.doTick(WorldServer.java:497)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.MinecraftServer.b(MinecraftServer.java:1245)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.DedicatedServer.b(DedicatedServer.java:430)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.MinecraftServer.a(MinecraftServer.java:1112)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.v1_15_R1.MinecraftServer.run(MinecraftServer.java:934)
[04:55:48] [Paper Watchdog Thread/ERROR]: 		java.lang.Thread.run(Unknown Source)
[04:55:48] [Paper Watchdog Thread/ERROR]: ------------------------------
[04:55:48] [Paper Watchdog Thread/ERROR]: --- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH ---
[04:55:48] [Paper Watchdog Thread/ERROR]: ------------------------------

**Comment by SrBedrock** [comment]
```
>.... [21:59:10 FATAL]: Error executing task on Server
java.lang.NullPointerException: null
at net.minecraft.server.v1_15_R1.ChunkMapDistance.b(ChunkMapDistance.java:230) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PlayerChunkMap.movePlayer(PlayerChunkMap.java:1409) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.ChunkProviderServer.movePlayer(ChunkProviderServer.java:811) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PlayerConnection.a(PlayerConnection.java:1145) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PacketPlayInFlying.a(SourceFile:122) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PacketPlayInFlying$PacketPlayInPosition.a(SourceFile:56) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PlayerConnectionUtils.lambda$ensureMainThread$0(PlayerConnectionUtils.java:23) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.TickTask.run(SourceFile:18) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.executeTask(IAsyncTaskHandler.java:136) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.executeNext(IAsyncTaskHandler.java:109) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.MinecraftServer.ba(MinecraftServer.java:1038) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.MinecraftServer.executeNext(MinecraftServer.java:1031) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.awaitTasks(IAsyncTaskHandler.java:119) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.MinecraftServer.a(MinecraftServer.java:1102) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.MinecraftServer.run(MinecraftServer.java:934) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
>.... [21:59:12 FATAL]: Error executing task on Server
java.lang.NullPointerException: null
at net.minecraft.server.v1_15_R1.ChunkMapDistance.b(ChunkMapDistance.java:230) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PlayerChunkMap.movePlayer(PlayerChunkMap.java:1409) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.ChunkProviderServer.movePlayer(ChunkProviderServer.java:811) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PlayerConnection.a(PlayerConnection.java:1145) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PacketPlayInFlying.a(SourceFile:122) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PacketPlayInFlying$PacketPlayInPosition.a(SourceFile:56) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PlayerConnectionUtils.lambda$ensureMainThread$0(PlayerConnectionUtils.java:23) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.TickTask.run(SourceFile:18) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.executeTask(IAsyncTaskHandler.java:136) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.executeNext(IAsyncTaskHandler.java:109) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.MinecraftServer.ba(MinecraftServer.java:1038) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.MinecraftServer.executeNext(MinecraftServer.java:1031) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.awaitTasks(IAsyncTaskHandler.java:119) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.MinecraftServer.a(MinecraftServer.java:1102) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.MinecraftServer.run(MinecraftServer.java:934) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
>.... [21:59:12 FATAL]: Error executing task on Server
java.lang.NullPointerException: null
at net.minecraft.server.v1_15_R1.ChunkMapDistance.b(ChunkMapDistance.java:230) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PlayerChunkMap.movePlayer(PlayerChunkMap.java:1409) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.ChunkProviderServer.movePlayer(ChunkProviderServer.java:811) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PlayerConnection.a(PlayerConnection.java:1145) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PacketPlayInFlying.a(SourceFile:122) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PacketPlayInFlying$PacketPlayInPosition.a(SourceFile:56) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.PlayerConnectionUtils.lambda$ensureMainThread$0(PlayerConnectionUtils.java:23) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.TickTask.run(SourceFile:18) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.executeTask(IAsyncTaskHandler.java:136) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.executeNext(IAsyncTaskHandler.java:109) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.MinecraftServer.ba(MinecraftServer.java:1038) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.MinecraftServer.executeNext(MinecraftServer.java:1031) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.IAsyncTaskHandler.awaitTasks(IAsyncTaskHandler.java:119) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.MinecraftServer.a(MinecraftServer.java:1102) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at net.minecraft.server.v1_15_R1.MinecraftServer.run(MinecraftServer.java:934) ~[patched_1.15.2.jar:git-Paper-"9f5fadcc"]
at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
```

**Comment by froobynooby** [comment]
That first one is saying your server took too long to tick, doesn't look related to what this patch added though

The second looks like #2803

**Comment by SrBedrock** [comment]
New error/crash:

- Full log: https://mclo.gs/voto5xh
- Error: 
  - https://hasteb.in/xidayocu.cs
  - https://hasteb.in/ehihuket.css
- Crash Report: https://hasteb.in/ovusixiy.php

**Comment by froobynooby** [comment]
Are you using the updated paperclip? Still says you're using `git-Paper-"49495b62"`, which is the old version

**Comment by SrBedrock** [comment]
Oops, I was using the old jar.
sorry, after the errors i updated the jar.

**Comment by SrBedrock** [comment]
So far no errors have appeared on the console.

**Comment by TNTUP** [comment]
That would solve those "EntityTick" I see on several timings? I shall try this morning

**Comment by HexedHero** [comment]
> That would solve those "EntityTick" I see on several timings? I shall try this morning

I highly recommend testing

**Comment by TNTUP** [comment]
> > That would solve those "EntityTick" I see on several timings? I shall try this morning
> 
> I highly recommend testing

Will do this morning, when all players are sleeping :P I hate to disturb them just for testing purposes, and doing a backup just in case! ^^

**Comment by froobynooby** [comment]
It won't help with `entityTick` times, sorry. It only helps with packet processing times for movement packets - so if you see stuff like `Minecraft::Packet - PacketPlayInPositionLook` high in your timings reports it may help in reducing that

**Comment by Fejm** [comment]
> That would solve those "EntityTick" I see on several timings? I shall try this morning

Just limit mobs on map. Spigot/Paper can handle ~2.5k mobs + 100 players with TPS 20. Any higher numbers will cause lag.

**Comment by SrBedrock** [comment]
@froobynooby could you make the patch available in the latest version?

File: Spigot-Server-Patches/0440-Reduce-entity-tracker-updates-on-move.patch [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+index 7d408542e..2ae44b230 100644
+--- a/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
++++ b/src/main/java/com/destroystokyo/paper/PaperWorldConfig.java
+@@ -658,4 +658,9 @@ public class PaperWorldConfig {
+     private void nerfNetherPortalPigmen() {
+         nerfNetherPortalPigmen = getBoolean("game-mechanics.nerf-pigmen-from-nether-portals", nerfNetherPortalPigmen);
+     }
++
++    public double trackerUpdateDistance = 1;
++    private void trackeruUpdateDistance() {
```
Comment (HexedHero) [inline Spigot-Server-Patches/0440-Reduce-entity-tracker-updates-on-move.patch:52]: Spelling mistake?
trackerUUpdate

**Comment by Fabrimat** [comment]
Any news?

**Comment by Fabrimat** [comment]
Tested for 24h on a Survival with an average of 40 players. No errors in console or crashes, right now with 50 players there are about 15 TPS (before the patch 10 TPS).

**Comment by SrBedrock** [comment]
Any update @froobynooby ?

**Comment by paul-maxime** [comment]
I'm using this patch on my survival server (about 30 players online) and the TPS definitely went up.
I'll report if I come across any crash!

**Comment by iOshawott** [comment]
Definitely a patch, which is needed, I'm using it and waiting for updates

**Comment by rayanbzd** [comment]
This patch is really cool. 
Can u update on paper 132 ? For test it with latest update ? 
i'm running on a server with 80 players onlines on survival.

**Comment by clrxbl** [comment]
Here's my updated Paper-140 build, use at own risk.

https://cdn.discordapp.com/attachments/660148702224318471/689825524905803845/paperclip.jar

**Comment by rayanbzd** [comment]
Insane, I went from 12TPS (without the patch) to 16-17TPS (with the patch) with 110 players online. Server hosting >90 simultaneous players in 1.15.2 every days

None of the players have noticed any bugs.

However this does not totally replace EntityTrackerFixer, and it is important to keep it as a complement for servers with more than 40 players. 

It is becoming urgent for many servers to implement this patch. 
Thanks again for the great work

**Comment by rayanbzd** [comment]
Can someone build this patch with paper-149 based please ? :)

**Comment by LoXR** [comment]
Pretty sure Tuinity has this merged now?

**Review by electronicboy — APPROVED** [review]
Fixed some (minor/nitpick) concerns locally

**Comment by electronicboy** [comment]
Sorry for the delays, merged manually
https://github.com/PaperMC/Paper/commit/6bf04cd5ed11b3dd8e561da8d8cbe4ff8926fd8c
